### ISSUE TITLE
feature: retry with backoff if rook operator fails to create csi resources

### ISSUE THREAD

**Issue by adabuleanu** [rook/rook#7483]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Rook operator did not create the csi resources

**Expected behavior:**
Rook operator should have created the csi resources

**How to reproduce it (minimal and precise):**
This does not reproduce every time. I reproduced it in some of our pipelines

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
See attached file: [ceph-cluster.log](https://github.com/rook/rook/files/6210886/ceph-cluster.log)
* Operator's logs, if necessary
See attached file: [ceph-operator.log](https://github.com/rook/rook/files/6210888/ceph-operator.log)
* Crashing pod(s) logs, if necessary
See attached file: [pod.log](https://github.com/rook/rook/files/6210889/pod.log)

**Environment**:
* OS (e.g. from /etc/os-release): CentOS Linux 7 (Core)
* Kernel (e.g. `uname -a`):  3.10.0-1127.el7.x86_64
* Cloud provider or hardware configuration: Openstack
* Rook version (use `rook version` inside of a Rook Pod): v1.5.0-alpha.0.608.g3f128a4
* Storage backend version (e.g. for ceph do `ceph -v`): 15.2.9 (357616cbf726abb779ca75a551e8d02568e15b17) octopus (stable)
* Kubernetes version (use `kubectl version`): v1.20.4
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Vanilla Kubernetes
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):
```
cluster:
    id:     d1be64c3-e038-449b-b8da-45aef81564ac
    health: HEALTH_WARN
            noout flag(s) set
            2 daemons have recently crashed
 
  services:
    mon: 3 daemons, quorum a,b,c (age 3h)
    mgr: a(active, since 23h)
    mds: rook-ceph-ceph-fs:1 {0=rook-ceph-ceph-fs-a=up:active} 1 up:standby-replay
    osd: 3 osds: 3 up (since 14h), 3 in (since 23h)
         flags noout
 
  data:
    pools:   4 pools, 97 pgs
    objects: 22 objects, 16 KiB
    usage:   3.0 GiB used, 57 GiB / 60 GiB avail
    pgs:     97 active+clean
 
  io:
    client:   852 B/s rd, 1 op/s rd, 0 op/s wr
```

**Comment by leseb** [comment]
The error:

```
2021-03-25 10:45:00.755870 E | ceph-csi: failed to start Ceph csi drivers. failed to start rbdplugin daemonset: &DaemonSet{ObjectMeta:{csi-rbdplugin  kube-system    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [{apps/v1 Deployment rook-ceph-operator 505bd2cb-9c16-40c8-96d4-3cec1f7afa07 0xc0006f1317 0xc00096a9d9}] []  []},Spec:DaemonSetSpec{Selector:&v1.LabelSelector{MatchLabels:map[string]string{app: csi-rbdplugin,},MatchExpressions:[]LabelSelectorRequirement{},},Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[app:csi-rbdplugin contains:csi-rbdplugin-metrics] map[] [] []  []} {[{plugin-dir {HostPathVolumeSource{Path:/var/lib/kubelet/plugins/kube-system.rbd.csi.ceph.com,Type:*DirectoryOrCreate,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {plugin-mount-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/plugins,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {registration-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/plugins_registry/,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {pods-mount-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/pods,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-dev {&HostPathVolumeSource{Path:/dev,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-sys {&HostPathVolumeSource{Path:/sys,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {lib-modules {&HostPathVolumeSource{Path:/lib/modules,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {ceph-csi-config {nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil &ConfigMapVolumeSource{LocalObjectReference:LocalObjectReference{Name:rook-ceph-csi-config,},Items:[]KeyToPath{KeyToPath{Key:csi-cluster-config-json,Path:config.json,Mode:nil,},},DefaultMode:nil,Optional:nil,} nil nil nil nil nil nil nil nil nil nil}} {keys-tmp-dir {nil &EmptyDirVolumeSource{Medium:Memory,SizeLimit:<nil>,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-run-mount {&HostPathVolumeSource{Path:/run/mount,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}}] [] [{driver-registrar nci-docker.artifactory3.ds.jdsu.net/k8scsi/csi-node-driver-registrar:v2.0.1 [] [--v=0 --csi-address=/csi/csi.sock --kubelet-registration-path=/var/lib/kubelet/plugins/kube-system.rbd.csi.ceph.com/csi.sock]  [] [] [{KUBE_NODE_NAME  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] {map[] map[]} [{plugin-dir false /csi  <nil> } {registration-dir false /registration  <nil> }] [] nil nil nil nil    &SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false} {csi-rbdplugin nci-docker.artifactory3.ds.jdsu.net/cephcsi/cephcsi:v3.2.0 [] [--nodeid=$(NODE_ID) --endpoint=$(CSI_ENDPOINT) --v=0 --type=rbd --nodeserver=true --drivername=kube-system.rbd.csi.ceph.com --pidlimit=-1 --metricsport=9090 --metricspath=/metrics --enablegrpcmetrics=true]  [] [] [{POD_IP  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:status.podIP,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {NODE_ID  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {POD_NAMESPACE  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {CSI_ENDPOINT unix:///csi/csi.sock nil}] {map[] map[]} [{plugin-dir false /csi  <nil> } {pods-mount-dir false /var/lib/kubelet/pods  0xc000fb5cc0 } {plugin-mount-dir false /var/lib/kubelet/plugins  0xc000fb5cd0 } {host-dev false /dev  <nil> } {host-sys false /sys  <nil> } {lib-modules true /lib/modules  <nil> } {ceph-csi-config false /etc/ceph-csi-config/  <nil> } {keys-tmp-dir false /tmp/csi/keys  <nil> } {host-run-mount false /run/mount  <nil> }] [] nil nil nil nil   IfNotPresent &SecurityContext{Capabilities:&Capabilities{Add:[SYS_ADMIN],Drop:[],},Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:*true,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false} {liveness-prometheus nci-docker.artifactory3.ds.jdsu.net/cephcsi/cephcsi:v3.2.0 [] [--type=liveness --endpoint=$(CSI_ENDPOINT) --metricsport=9080 --metricspath=/metrics --polltime=60s --timeout=3s]  [] [] [{CSI_ENDPOINT unix:///csi/csi.sock nil} {POD_IP  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:status.podIP,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] {map[] map[]} [{plugin-dir false /csi  <nil> }] [] nil nil nil nil   IfNotPresent &SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false}] []  <nil> <nil> ClusterFirstWithHostNet map[]  rook-csi-rbd-plugin-sa <nil>  true true false <nil> nil []   &Affinity{NodeAffinity:&NodeAffinity{RequiredDuringSchedulingIgnoredDuringExecution:nil,PreferredDuringSchedulingIgnoredDuringExecution:[]PreferredSchedulingTerm{},},PodAffinity:nil,PodAntiAffinity:nil,}  [{ Exists  NoSchedule <nil>}] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil>}},UpdateStrategy:DaemonSetUpdateStrategy{Type:OnDelete,RollingUpdate:nil,},MinReadySeconds:0,RevisionHistoryLimit:nil,},Status:DaemonSetStatus{CurrentNumberScheduled:0,NumberMisscheduled:0,DesiredNumberScheduled:0,NumberReady:0,ObservedGeneration:0,UpdatedNumberScheduled:0,NumberAvailable:0,NumberUnavailable:0,CollisionCount:nil,Conditions:[]DaemonSetCondition{},},}: failed to start csi-rbdplugin daemonset: Internal error occurred: resource quota evaluation timed out
&DaemonSet{ObjectMeta:{csi-rbdplugin  kube-system    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [{apps/v1 Deployment rook-ceph-operator 505bd2cb-9c16-40c8-96d4-3cec1f7afa07 0xc0006f1317 0xc00096a9d9}] []  []},Spec:DaemonSetSpec{Selector:&v1.LabelSelector{MatchLabels:map[string]string{app: csi-rbdplugin,},MatchExpressions:[]LabelSelectorRequirement{},},Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[app:csi-rbdplugin contains:csi-rbdplugin-metrics] map[] [] []  []} {[{plugin-dir {HostPathVolumeSource{Path:/var/lib/kubelet/plugins/kube-system.rbd.csi.ceph.com,Type:*DirectoryOrCreate,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {plugin-mount-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/plugins,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {registration-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/plugins_registry/,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {pods-mount-dir {&HostPathVolumeSource{Path:/var/lib/kubelet/pods,Type:*Directory,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-dev {&HostPathVolumeSource{Path:/dev,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-sys {&HostPathVolumeSource{Path:/sys,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {lib-modules {&HostPathVolumeSource{Path:/lib/modules,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {ceph-csi-config {nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil &ConfigMapVolumeSource{LocalObjectReference:LocalObjectReference{Name:rook-ceph-csi-config,},Items:[]KeyToPath{KeyToPath{Key:csi-cluster-config-json,Path:config.json,Mode:nil,},},DefaultMode:nil,Optional:nil,} nil nil nil nil nil nil nil nil nil nil}} {keys-tmp-dir {nil &EmptyDirVolumeSource{Medium:Memory,SizeLimit:<nil>,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}} {host-run-mount {&HostPathVolumeSource{Path:/run/mount,Type:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}}] [] [{driver-registrar nci-docker.artifactory3.ds.jdsu.net/k8scsi/csi-node-driver-registrar:v2.0.1 [] [--v=0 --csi-address=/csi/csi.sock --kubelet-registration-path=/var/lib/kubelet/plugins/kube-system.rbd.csi.ceph.com/csi.sock]  [] [] [{KUBE_NODE_NAME  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] {map[] map[]} [{plugin-dir false /csi  <nil> } {registration-dir false /registration  <nil> }] [] nil nil nil nil    &SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false} {csi-rbdplugin nci-docker.artifactory3.ds.jdsu.net/cephcsi/cephcsi:v3.2.0 [] [--nodeid=$(NODE_ID) --endpoint=$(CSI_ENDPOINT) --v=0 --type=rbd --nodeserver=true --drivername=kube-system.rbd.csi.ceph.com --pidlimit=-1 --metricsport=9090 --metricspath=/metrics --enablegrpcmetrics=true]  [] [] [{POD_IP  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:status.podIP,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {NODE_ID  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {POD_NAMESPACE  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {CSI_ENDPOINT unix:///csi/csi.sock nil}] {map[] map[]} [{plugin-dir false /csi  <nil> } {pods-mount-dir false /var/lib/kubelet/pods  0xc000fb5cc0 } {plugin-mount-dir false /var/lib/kubelet/plugins  0xc000fb5cd0 } {host-dev false /dev  <nil> } {host-sys false /sys  <nil> } {lib-modules true /lib/modules  <nil> } {ceph-csi-config false /etc/ceph-csi-config/  <nil> } {keys-tmp-dir false /tmp/csi/keys  <nil> } {host-run-mount false /run/mount  <nil> }] [] nil nil nil nil   IfNotPresent &SecurityContext{Capabilities:&Capabilities{Add:[SYS_ADMIN],Drop:[],},Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:*true,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false} {liveness-prometheus nci-docker.artifactory3.ds.jdsu.net/cephcsi/cephcsi:v3.2.0 [] [--type=liveness --endpoint=$(CSI_ENDPOINT) --metricsport=9080 --metricspath=/metrics --polltime=60s --timeout=3s]  [] [] [{CSI_ENDPOINT unix:///csi/csi.sock nil} {POD_IP  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:,FieldPath:status.podIP,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] {map[] map[]} [{plugin-dir false /csi  <nil> }] [] nil nil nil nil   IfNotPresent &SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,} false false false}] []  <nil> <nil> ClusterFirstWithHostNet map[]  rook-csi-rbd-plugin-sa <nil>  true true false <nil> nil []   &Affinity{NodeAffinity:&NodeAffinity{RequiredDuringSchedulingIgnoredDuringExecution:nil,PreferredDuringSchedulingIgnoredDuringExecution:[]PreferredSchedulingTerm{},},PodAffinity:nil,PodAntiAffinity:nil,}  [{ Exists  NoSchedule <nil>}] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil>}},UpdateStrategy:DaemonSetUpdateStrategy{Type:OnDelete,RollingUpdate:nil,},MinReadySeconds:0,RevisionHistoryLimit:nil,},Status:DaemonSetStatus{CurrentNumberScheduled:0,NumberMisscheduled:0,DesiredNumberScheduled:0,NumberReady:0,ObservedGeneration:0,UpdatedNumberScheduled:0,NumberAvailable:0,NumberUnavailable:0,CollisionCount:nil,Conditions:[]DaemonSetCondition{},},}
```

Probably this `failed to start csi-rbdplugin daemonset: Internal error occurred: resource quota evaluation timed out`. Looks more like a Kubernetes issue at this point.

**Comment by adabuleanu** [comment]
In our deployment we do not alter k8s in any way before installing rook. It may be the case that there is an etcd timeout issue. Shouldn't the rook operator cope with this kind of issues, like using a retry mechanism?

**Comment by leseb** [comment]
We could retry with backoff indeed, we just don't at the moment. It's the first time we are seeing this.

**Comment by travisn** [comment]
This should be fixed by #8020 in v1.6.5
